server/src/main/java/org/elasticsearch/search/aggregations/metrics/max/MaxAggregatorFactory.java

```diff
@@ -43,13 +53,71 @@ public MaxAggregatorFactory(String name, ValuesSourceConfig<Numeric> config, Sea
     @Override
     protected Aggregator createUnmapped(Aggregator parent,
             List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
-        return new MaxAggregator(name, null, config.format(), context, parent, pipelineAggregators, metaData);
+        return new MaxAggregator(name, null, config.format(), context, parent, pipelineAggregators, metaData, null);
     }
 
     @Override
     protected Aggregator doCreateInternal(ValuesSource.Numeric valuesSource, Aggregator parent,
             boolean collectsFromSingleBucket, List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData)
                     throws IOException {
-        return new MaxAggregator(name, valuesSource, config.format(), context, parent, pipelineAggregators, metaData);
+        final Function<byte[], Number> converter = getPointReaderOrNull(context, getParent(), config);
+        final CheckedFunction<LeafReader, Number, IOException> shortcutMaxFunc;
+        if (converter == null) {
+            shortcutMaxFunc = createEmpty();
+        } else {
+            final String fieldName = config.fieldContext().field();
+            shortcutMaxFunc = createShortcutMax(fieldName, converter);
+        }
+        return new MaxAggregator(name, valuesSource, config.format(), context, parent, pipelineAggregators, metaData, shortcutMaxFunc);
+    }
+
+    /**
+     * Returns a function that can extract the maximum value of a field using
+     * {@link PointValues}.
+     * @param fieldName The name of the field.
+     * @param converter The point value converter.
+     */
+    public static CheckedFunction<LeafReader, Number, IOException> createShortcutMax(String fieldName, Function<byte[], Number> converter) {
```

Comment:
This feels weird to me here. I wonder if we should instead have this method in the aggregator class and pass in a boolean parameter that determines whether we use it?

Reply:
We need the `converter` so we could pass it to the aggregator directly and if it's not null we can try to extract the min/max value from points ?

Reply:
++ I like that better. I'm not a fan of the Aggregators calling back out to the factory so if we can help that not happen then I'm on board 😄

Reply:
can it be pkg-private?

Reply:
This is used in tests that are in a different package (`metrics`). We could probably merge all `metrics.*` in a single package instead of having one package per metrics agg ?